feat(ui): Add image gallery component

saranicoly opened the pull request:
## What's the purpose of this pull request?
<em>The [image gallery](https://www.figma.com/file/zHMOtOPduiM0hNNcg21iR1/Handoff-PDP-v0.1?node-id=18%3A33225) component shows all the product images on PDP.

<img width="925" alt="Screen Shot 2021-12-23 at 21 16 49" src="https://user-images.githubusercontent.com/24229855/147301290-bc1c2ef3-3616-4058-8852-c526b7061b7f.png">

It's currently hidden from PDP because it won't be used right now.
</em>

## How it works? 
<em>To use ImageGallery, you must import it from `src/components/ui/ImageGallery` and call the component like this:
`<ImageGallery images={imagesArray}/>`

**imagesArray** is an array of `ImageElementData`:
```TypeScript
export interface ImageElementData {
  url: string
  alternateName: string
}
```
Please note that using Thumbor generated images will not work as expected. Use simple image urls if you want to create this array of objects manually. Good image sources: 
- [Placekitten](https://placekitten.com/)
- [Pexels](https://www.pexels.com/)
</em>

## How to test it?
<em> 

- Import ImageGallery in `ProductDetails.tsx` and change Image tag for: `<ImageGallery images={productImages} />`. You can also hardcode your own array of `ImageElementData` using your own image urls.
- run `yarn develop` and `yarn test`
</em>

saranicoly committed:
add basic structure

saranicoly committed:
add basic structure

saranicoly committed:
fix typo

saranicoly committed:
fix typo

saranicoly committed:
add image gallery to product details

saranicoly committed:
add image gallery to product details

saranicoly committed:
using carrossel

saranicoly committed:
using carrossel

saranicoly committed:
add CarouselSlider

saranicoly committed:
add CarouselSlider

Commit:
Add initial ImageGallery prototype

Commit:
Add initial ImageGallery prototype

saranicoly committed:
fix error

saranicoly committed:
fix error

Commit:
Change ImageGallerySelector implementation

Commit:
Change ImageGallerySelector implementation

saranicoly committed:
renaming CarouselSlider

saranicoly committed:
Co-authored-by: Heitor Carvalho <heitor.carvalho@vtex.com.br>

saranicoly committed:
remove carouselSample

saranicoly committed:
add onClick

saranicoly committed:
working selector wip

saranicoly committed:
improve imageGallerySelector

saranicoly committed:
setting the first element as default

Commit:
Integrate ImageGallery with data from GQL

Co-authored-by: Sara Nicoly <saranicoly0@gmail.com>

Commit:
Remove styling from inner button

Commit:
Add interface to pass src and alt image attributes simultaneously

Commit:
Remove ImageGallery from ProductDetails

Add <Image> component back for alpha release

Commit:
Cleanup, remove more styles, and use Image component on ImageGalleryZoom

saranicoly committed:
change arrow names and add aria-label for both

Commit:
Add data attribute and dynamic aria-label to thumbnail button

Commit:
Add selected class to active thumbnail button

Co-authored-by: Sara Nicoly <saranicoly0@gmail.com>

saranicoly committed:
Merge branch 'master' into feat/image-gallery

israelswf commented:
Just a minor suggestion I've notice now when following the _How to_ steps ... change:
`<ImageGallery images={all_images}/>`
to
`<ImageGallery images={allImages}/>`

Eslint is complaining about the `all_images` variable name in _snake case_.

igorbrasileiro commented on src/components/ui/ImageGallery/ImageGalleryZoom.tsx:
```diff
@@ -0,0 +1,15 @@
+import type { PropsWithChildren } from 'react'
+import React from 'react'
+
+interface ImageGalleryZoomProps {
+  helpMessage?: string
+  zoomFactor?: number
+}
+
+const ImageGalleryZoom = ({
+  children,
+}: PropsWithChildren<ImageGalleryZoomProps>) => {
+  return <> {children} </>
+}
+
+export default ImageGalleryZoom
```
Remove this noop component

saranicoly replied:
The idea of having this component is just to be a placeholder with some basic structure, so it’s going to be ready to receive its functionality later. With that in mind, we think this should be in a separated file like this because it's not being used right now but it will in the future.

saranicoly replied:
Also, the zoom component has value itself. So maybe it could be reused in other structures.

igorbrasileiro replied:
Ok, no problem.

igorbrasileiro reviewed (changes requested):
Great job, @saranicoly! I left some comments.
I have a few questions, why need to reimplement the whole carousel?
Also, what do you think about this structure?

```tsx
const ImageGallery = ({ images }) => {
   const [selectedImage, setSelectedImage] = useState(0)
   const image = images[selectedImage]
   
   <div>
      <ZoomImage>
         <Image {...image} />
      </ZoomImage>
      <Carousel>
        {images.map((image, index) => {
          return <Image {...image} onClick={(e) => { setSelectedImage(index) })
        })}
      </Carousel>
   </div>
}
```

Maybe I'm forgetting something, but IMO it will make this component a little bit easier to understand.

lariciamota commented:
Hi! I wasn't able to open the figma link :/
<img width="490" alt="Screen Shot 2021-12-22 at 16 52 24" src="https://user-images.githubusercontent.com/19983991/147147792-cb1c8ac2-ac04-40e4-80b1-94c8ad547074.png">

saranicoly committed:
change icon location inside iconButton

saranicoly commented on src/components/ui/ImageGallery/ImageGallerySelector.tsx:
```diff
@@ -0,0 +1,139 @@
+          slide('next', sliderDispatch)
+        }}
+        icon={<ForwardArrowIcon color="#323845" />}
+      />
```
Done. Thanks 😊

heitorado replied:
Fixed at https://github.com/vtex-sites/base.store/pull/142/commits/a9e1b45dfc591c5d5364d46eda79c9f6094fb61c

Commit:
Rewrite ImageGallery to make it simpler

- Removed the compound components pattern (provider and useImageGallery hook)
- Moved the internal selector logic to the main ImageGallery
- The state now stores only the image data index

Co-authored-by: Sara Nicoly <saranicoly0@gmail.com>

heitorado commented on src/components/ui/ImageGallery/ImageGalleryProvider.tsx:
```diff
@@ -0,0 +1,27 @@
+
+  return (
+    <ImageGalleryContext.Provider value={contextValues}>
+      <div {...otherProps}>{children}</div>
```
We adopted your previous suggestion of structure, so `ImageGalleryContext.Provider` does not exist anymore.

heitorado commented on src/components/ui/ImageGallery/ImageGallery.tsx:
```diff
@@ -0,0 +1,48 @@
+              target.getAttribute('alt') ?? previousSelectedImage.alternateName,
+          }
+        })
+      }
```
This initial design was putting the `onClick` in the button so the `e.target` would be the image component, from where we accessed the `src` and `alt` parameters. Now after the rewrite on the structure, the `onClick` is on the button that wraps the image component and does not need to grab attributes from its child anymore. The main reason that we did not put the `onClick` directly into the image, but on the Button instead, is to give better semantics.

We thought about adding the `button` role, but we found [here](https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/button_role) that some extra work is necessary to make it _behave_ like a proper button. Thus, we decided to stick with the Button component instead.

heitorado commented on src/components/ui/ImageGallery/ImageGallerySelector.tsx:
```diff
@@ -0,0 +1,139 @@
+  return transformMap
+}
+
+function ImageGallerySelector({ imagesPerPage, images, ...otherProps }: Props) {
```
I agree. This is, in fact, an adaptation of the shelf implementation made by @georgebrindeiro to the faststore 101 class. We thought about renaming it to "Slider" but we already have a component with that name. Any thoughts/suggestions on this?

saranicoly replied:
What do you think about `ImageSelector` or just `Selector`?

igorbrasileiro replied:
I prefer to name it a carousel because the slider is a kind of input.

Commit:
Move selector icons to appear before the other selector elements

Co-authored-by: Sara Nicoly <saranicoly0@gmail.com>

heitorado commented:
> I have a few questions, why need to reimplement the whole carousel?

Hmm, it wasn't our intention to reimplement the carousel. In fact, initially we tried to use the Carousel component, but we perceived that:
- It isn't flexible about the number of elements to be displayed on a page
- We tried to update the component to add that flexibility, but apparently there is a lot of internal logic that relies on displaying a single item per page. We found that it was easier to work with the basic shelf implementation than the Carousel.
- Also, we would have to add the flexibility to receive custom arrow icons and the amount of items per page.

Commit:
Rename ImageGalleryZoom to ImageZoom

Co-authored-by: Sara Nicoly <saranicoly0@gmail.com>

saranicoly commented:
> Hi! I wasn't able to open the figma link :/ <img alt="Screen Shot 2021-12-22 at 16 52 24" width="490" src="https://user-images.githubusercontent.com/19983991/147147792-cb1c8ac2-ac04-40e4-80b1-94c8ad547074.png">

Sorry, I think you don't have access. 😢 
We will change the PR description and put an image instead of the Figma link so that everyone can see it.

saranicoly committed:
Merge branch 'master' into feat/image-gallery

igorbrasileiro reviewed (approved):
Great job, @saranicoly, and @heitorado!

eduardoformiga commented on src/components/ui/ImageGallery/ImageGallerySelector.tsx:
```diff
@@ -0,0 +1,130 @@
+  }
+
+  const transformMap: Record<number, number> = {}
+  const slideWidth = 100 / totalItems
```
What does this 100 magic number represent? the entire width?

heitorado replied:
Yes, it is the entire width of the element on which the ImageGallery is nested. I'll add a comment to make that explicit.

heitorado replied:
Done at https://github.com/vtex-sites/base.store/pull/142/commits/8f303c5e398e46c7c0396bafbdd32aec813322aa

eduardoformiga commented on src/components/ui/ImageGallery/ImageGallerySelector.tsx:
```diff
@@ -0,0 +1,130 @@
+
+  const slidingTransition = `transform 400ms`
+
+  if (!elements || !elementCount) {
```
```suggestion
  if (!elementsCount) {
```
I think the first condition (![]) is always false since the array is an object and is always evaluated as a [truthy](https://developer.mozilla.org/pt-BR/docs/Glossary/Truthy) value, but with the `!` operator will be always `falsy`. 

I also wonder if we should move this guard clause to the beginning of the function.

heitorado replied:
If we try to move the guard clause to the beginning of the function, this error is reported from the linter:

> _React Hook "useSlider" is called conditionally. React Hooks must be called in the exact same order in every component render. Did you accidentally call a React Hook after an early return?_
(similar error arises for `useMemo`)

https://reactjs.org/docs/hooks-rules.html

heitorado replied:
About that:

> I think the first condition (![]) is always false since the array is an object and is always evaluated as a [truthy](https://developer.mozilla.org/pt-BR/docs/Glossary/Truthy) value, but with the `!` operator will be always `falsy`.

You are right. I thought that if the children happened to be `[]`, `null` or `undefined`, the `.toArray` method would behave unexpectedly, but it always returns an empty array. And like you said, `![]` is always `false`. Fixed at

eduardoformiga commented on src/components/ui/ImageGallery/ImageGallerySelector.tsx:
```diff
@@ -0,0 +1,130 @@
+            display: 'flex',
+            transition: sliderState.sliding ? slidingTransition : undefined,
+            width: `${(elementCount * 100) / itemsPerPage}%`,
+            transform: `translate3d(${
```
🆒 !

heitorado replied:
Credits to @georgebrindeiro

igorbrasileiro replied:
Something interesting, why not calculate only the current transformValue instead of all values? I am asking because I wanna remove the useMemo, if it's possible.

heitorado replied:
Good call. I've removed the `useMemo` here: https://github.com/vtex-sites/base.store/pull/142/commits/d244c4dec6b48a0c4a97f8f7379f88285bd8739d

eduardoformiga commented on src/components/ui/ImageGallery/ImageGallerySelector.tsx:
```diff
@@ -0,0 +1,130 @@
+        <div
+          data-carousel-track
+          style={{
+            display: 'flex',
```
Nice job here!
I think you have a plan to move the static CSS (the ones that don't depend on js values) in the style PR afterward, right?

saranicoly replied:
Yes, I think so.

lariciamota replied:
I'm curious, why will it be done separately?

heitorado replied:
The ImageGallerySelector will not be present in the alpha version of the styled base.store. Since we are focusing on different component integrations/styling now, the styling for this component will be on hold for a while.

eduardoformiga reviewed (commented):
Amazing job, folks! ✨  
I Left some comments.

heitorado commented on src/components/sections/ProductDetails/ProductDetails.tsx:
```diff
@@ -30,7 +30,7 @@ function ProductDetails({ product: staleProduct }: Props) {
       brand: { name: brandName },
       isVariantOf: { name, productGroupID: productId },
-      image: [img],
+      image: productImages,
```
When changing this, my initial thought was that it would be "one less thing to do" when we integrate the gallery to the PDP. Just pass the "productImages" to the component and it's done.

heitorado commented on src/components/ui/ImageGallery/ImageGallery.tsx:
```diff
@@ -0,0 +1,58 @@
+              key={idx}
+              data-thumbnail-button
+              className={image.url === currentImage.url ? 'selected' : ''}
+              aria-label={`${image.alternateName} - Image ${idx + 1} of ${
```
I'm not sure if we should completely remove it. One is for the image being currently displayed, and other is for the images in the selector. Since we have two different "contexts", I guess it would make sense for the screen reader to describe the image that is currently appearing and the list of images that can be clicked to switch the main image? I do not know much about screen readers as well.

I used this site as a reference:
https://www.rei.com/product/194789/teva-reember-slip-ons-womens

It is listed on [baymard](https://baymard.com/blog/accessibility-benchmark-launch) as having good accessibility (just saying this to explain that it is not a random store):

Also just noticed that I made a mistake - the main (selected) image does not have the second part of the alt (after the `-`). Maybe it would be better if we strip the "- Image x of y" part from the selected image, as it is in the aforementioned store?

eduardoformiga replied:
Good references 😄 

I said that because we have the `alternateName` in three different places:
- In the `Image` inside the `ImageZoom` as `alt`.
- In the `Button` inside `ImageGallerySelector` as `aria-label`.
- In the `Image` inside the `Button` as `alt`.

It could sound a little annoying for screen readers. But no worries, good job worrying about it!

heitorado replied:
Hmm got it. Just now I realized that we have it three times, in which two of these times is the same message. Not sure how can we improve it without sacrificing one `alt` or `aria-label`. Suggestions? @saranicoly, thoughts on this?

Commit:
Load only first image eager and the others lazily

Commit:
Inline sliding transition string

Commit:
Add aria-roledescription and aria-label to ImageGallerySelector section

Commit:
Removed neutral element from the conditional

(it would always evaluate to false)

heitorado commented on src/components/ui/ImageGallery/ImageGallerySelector.tsx:
```diff
@@ -0,0 +1,130 @@
+    infiniteMode: false,
+  })
+
+  const slidingTransition = `transform 400ms`
```
Done at https://github.com/vtex-sites/base.store/pull/142/commits/6507a1eebf04482dcb0ca2e91a24fd525a6effab

heitorado commented on src/components/ui/ImageGallery/ImageGallerySelector.tsx:
```diff
@@ -0,0 +1,130 @@
+  }
+
+  return (
+    <section>
```
Done at https://github.com/vtex-sites/base.store/pull/142/commits/d20395480eb39d162befb881bcf8add423298866

heitorado commented on src/components/ui/ImageGallery/ImageGallery.tsx:
```diff
@@ -0,0 +1,58 @@
+                src={image.url}
+                alt={image.alternateName}
+                variant="product.details"
+                loading="eager"
```
Good call. Done at https://github.com/vtex-sites/base.store/pull/142/commits/9091cff7757c18b78f22327472f751232117a1d6

heitorado commented on src/components/ui/ImageGallery/ImageGallery.tsx:
```diff
@@ -0,0 +1,58 @@
+          return (
+            <Button
+              key={idx}
+              data-thumbnail-button
+              className={image.url === currentImage.url ? 'selected' : ''}
```
Hmm, I don't know. I'd rather stay with the 'selected' in the classname rather than using the data attribute for that. Personally I'd prefer doing `[data-thumbnail-button].selected` instead of `[data-thumbnail-button='selected']`. On both cases, the 'default' button state would be styled with just `[data-thumbnail-button]`, right?

Not sure which way is better. @renatamottam, @saranicoly - thoughts on this?

hellofanny replied:
Let's go with igor's suggestion! Can you please use `[data-thumbnail-button='selected']` instead? We are trying to follow this pattern for the other components too :)

heitorado replied:
Great! So data attribute it is. Thanks, @igorbrasileiro

heitorado replied:
Done at https://github.com/vtex-sites/base.store/pull/142/commits/11880b496ee56279d428dca98d447ac477542b58

Commit:
Move 'selected' attribute from className to data attribute

lariciamota commented on src/components/ui/ImageGallery/ImageGallerySelector.tsx:
```diff
@@ -0,0 +1,128 @@
+  return (
+    <section aria-roledescription="carousel" aria-label="Product images">
+      <IconButton
+        aria-label="slide image selector backward"
+        icon={<BackwardArrowIcon color="#323845" />}
+        onClick={() => {
+          if (sliderState.sliding) {
+            return
+          }
+
+          slide('previous', sliderDispatch)
+        }}
+      />
+      <IconButton
+        aria-label="slide image selector forward"
```
What do you think about this suggestion?
I'm not 100% sure about the English, but for me putting it at the beginning sounds better and starts with the most important part of the component description
```suggestion
        aria-label="backward slide image selector"
        icon={<BackwardArrowIcon color="#323845" />}
        onClick={() => {
          if (sliderState.sliding) {
            return
          }

          slide('previous', sliderDispatch)
        }}
      />
      <IconButton
        aria-label="forward slide image selector"
```

heitorado replied:
Great suggestion. I believe the English is also fine.

igorbrasileiro commented on src/components/ui/ImageGallery/ImageZoom.tsx:
```diff
@@ -0,0 +1,13 @@
+}
+
+const ImageZoom = ({ children }: PropsWithChildren<ImageZoomProps>) => {
+  return <> {children} </>
```
```suggestion
  return children
```

heitorado replied:
I tried adding your change, but then I got the following error at `ImageGallery.tsx` when using the `<ImageZoom />`:

```
'ImageZoom' cannot be used as a JSX component.
  Its return type 'ReactNode' is not a valid JSX element.
    Type 'undefined' is not assignable to type 'Element | null'
```

heitorado committed:
Change aria-label text on buttons

Co-authored-by: Larícia Mota <laricia.mota@vtex.com.br>

Commit:
Update aria-label on gallery slider image button

Commit:
Add documentation to ImageGallerySelector

Commit:
Remove transform calculation from useMemo
